File: windows/WiFiDirectDemo.Protocol/ProtocolMessage.cs
using System;

namespace WiFiDirectDemo.Protocol;

public sealed class ProtocolMessage
{
    public string Type { get; set; } = string.Empty;

    public string Sender { get; set; } = string.Empty;

    public string Text { get; set; } = string.Empty;

    public DateTimeOffset TimestampUtc { get; set; } = DateTimeOffset.UtcNow;

    public static ProtocolMessage Hello(string sender, string text = "ready")
        => new ProtocolMessage
        {
            Type = "hello",
            Sender = sender,
            Text = text,
            TimestampUtc = DateTimeOffset.UtcNow,
        };

    public static ProtocolMessage Chat(string sender, string text)
        => new ProtocolMessage
        {
            Type = "chat",
            Sender = sender,
            Text = text,
            TimestampUtc = DateTimeOffset.UtcNow,
        };

    public static ProtocolMessage Ping(string sender, string text = "ping")
        => new ProtocolMessage
        {
            Type = "ping",
            Sender = sender,
            Text = text,
            TimestampUtc = DateTimeOffset.UtcNow,
        };
}

File: android/app/src/main/java/com/example/wifidirectdemo/ProtocolMessage.kt
package com.example.wifidirectdemo

import java.time.Instant
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

data class ProtocolMessage(
    val type: String,
    val sender: String,
    val text: String,
    val timestampUtc: String = Instant.now().toString()
) {
    fun toJsonLine(): String {
        return buildJsonObject {
            put("type", type)
            put("sender", sender)
            put("text", text)
            put("timestampUtc", timestampUtc)
        }.toString()
    }

    companion object {
        private val json = Json {
            ignoreUnknownKeys = true
        }

        fun hello(sender: String, text: String = "ready"): ProtocolMessage =
            ProtocolMessage("hello", sender, text)

        fun chat(sender: String, text: String): ProtocolMessage =
            ProtocolMessage("chat", sender, text)

        fun ping(sender: String, text: String = "ping"): ProtocolMessage =
            ProtocolMessage("ping", sender, text)

        fun fromJson(json: String): ProtocolMessage {
            val obj = this.json.parseToJsonElement(json).jsonObject
            return ProtocolMessage(
                type = obj["type"]?.jsonPrimitive?.content.orEmpty(),
                sender = obj["sender"]?.jsonPrimitive?.content.orEmpty(),
                text = obj["text"]?.jsonPrimitive?.content.orEmpty(),
                timestampUtc = obj["timestampUtc"]?.jsonPrimitive?.content.orEmpty()
            )
        }
    }
}
